File: cart-service/src/chaos-simulator.ts
/**
 * Chaos Engineering Simulator for safe infrastructure failure simulation.
 */

import * as os from 'os';
import * as fs from 'fs';
import * as path from 'path';
import { FeatureFlagClient } from './feature-flag-client';

interface ChaosConfig {
  // Disk Stress
  intensityLevel?: number;      // 1-10
  durationSeconds?: number;     // Duration in seconds
  
  // Pod Crash
  crashIntervalMinutes?: number;  // Interval between checks
  crashProbability?: number;      // 0-100 percentage
  
  // Network Delay
  delayMs?: number;              // Base delay in milliseconds
  jitterMs?: number;             // Random jitter ±ms
  
  // Allow any additional properties from feature flag config
  [key: string]: any;
}

interface SystemMetrics {
  timestamp: string;
  active_simulations: string[];
  cpu_usage_percent?: number;
  memory_usage_mb?: number;
}

/**
 * ChaosSimulator class for simulating infrastructure failures safely.
 */
class ChaosSimulator {
  private activeSimulations: Map<string, boolean> = new Map();
  private diskStressInterval?: NodeJS.Timeout;
  private crashSchedulerInterval?: NodeJS.Timeout;
  private tempFiles: string[] = [];

  constructor() {
    console.log('ChaosSimulator initialized');
  }

  /**
   * Check for active chaos flags and apply corresponding simulations.
   */
  async checkAndApplyChaos(
    serviceName: string,
    flagClient: FeatureFlagClient
  ): Promise<void> {
    try {
      // Check disk stress flag
      const diskStressFlag = await flagClient.getFlag('infrastructure_disk_stress');
      if (diskStressFlag?.enabled) {
        await this.simulateDiskStress(diskStressFlag.config);
      }

      // Check pod crash flag
      const podCrashFlag = await flagClient.getFlag('infrastructure_pod_crash');
      if (podCrashFlag?.enabled) {
        await this.simulatePodCrash(podCrashFlag.config, serviceName);
      }

      // Check network delay flag
      const networkDelayFlag = await flagClient.getFlag('infrastructure_network_delay');
      if (networkDelayFlag?.enabled) {
        await this.simulateNetworkDelay(networkDelayFlag.config);
      }
    } catch (error) {
      // Don't let chaos simulator errors crash the service
      console.error('Error in chaos simulation:', error);
    }
  }

  /**
   * Get current system metrics for monitoring chaos effects.
   */
  getSystemMetrics(): SystemMetrics {
    const metrics: SystemMetrics = {
      timestamp: new Date().toISOString(),
      active_simulations: Array.from(this.activeSimulations.entries())
        .filter(([_, active]) => active)
        .map(([name, _]) => name)
    };

    // Add basic memory metrics
    try {
      const memUsage = process.memoryUsage();
      metrics.memory_usage_mb = memUsage.rss / 1024 / 1024;
      
      // CPU usage (approximation using process.cpuUsage())
      const cpuUsage = process.cpuUsage();
      metrics.cpu_usage_percent = (cpuUsage.user + cpuUsage.system) / 1000000; // Convert to seconds
    } catch (error) {
      console.warn('Error collecting system metrics:', error);
    }

    return metrics;
  }

  /**
   * Simulate network latency between services.
   */
  private async simulateNetworkDelay(config: ChaosConfig): Promise<void> {
    const delayMs = config.delayMs ?? 2000;
    const jitterMs = config.jitterMs ?? 500;

    // Calculate actual delay with random jitter
    const jitter = Math.floor(Math.random() * (jitterMs * 2 + 1)) - jitterMs;
    const actualDelayMs = Math.max(0, delayMs + jitter);

    console.log(`🔥 CHAOS: Simulating network delay (${actualDelayMs}ms)`);

    // Apply the delay
    await new Promise(resolve => setTimeout(resolve, actualDelayMs));
  }

  /**
   * Simulate pod crashes by scheduling periodic exits.
   */
  private async simulatePodCrash(
    config: ChaosConfig,
    serviceName: string
  ): Promise<void> {
    // Check if already running
    if (this.activeSimulations.get('pod_crash')) {
      return;
    }

    const intervalMinutes = config.crashIntervalMinutes ?? 5;
    const crashProbability = config.crashProbability ?? 30;

    console.log(
      `🔥 CHAOS: Starting pod crash scheduler ` +
      `(interval=${intervalMinutes}min, probability=${crashProbability}%)`
    );
    this.activeSimulations.set('pod_crash', true);

    // Start crash scheduler
    const intervalMs = intervalMinutes * 60 * 1000;
    this.crashSchedulerInterval = setInterval(() => {
      // Check if we should crash based on probability
      const roll = Math.floor(Math.random() * 100) + 1;
      if (roll <= crashProbability) {
        console.error(
          `🔥 CHAOS: Simulating pod crash for ${serviceName} ` +
          `(probability=${crashProbability}%)`
        );
        // Force exit the process
        process.exit(1);
      }
    }, intervalMs);
  }

  /**
   * Simulate high disk I/O by creating intensive read/write operations.
   */
  private async simulateDiskStress(config: ChaosConfig): Promise<void> {
    // Check if already running
    if (this.activeSimulations.get('disk_stress')) {
      return;
    }

    const intensity = config.intensityLevel ?? 5;
    const duration = config.durationSeconds ?? 30;

    console.log(
      `🔥 CHAOS: Starting disk I/O stress simulation ` +
      `(intensity=${intensity}, duration=${duration}s)`
    );
    this.activeSimulations.set('disk_stress', true);

    // Start disk stress in background
    this.performDiskStress(intensity, duration);
  }

  /**
   * Perform disk stress operations in the background.
   */
  private performDiskStress(intensity: number, duration: number): void {
    const startTime = Date.now();
    const fileSizeMB = intensity; // 1-10 MB per file
    const numFiles = intensity; // 1-10 files
    const tmpDir = os.tmpdir();

    // Create temp files
    for (let i = 0; i < numFiles; i++) {
      const tempFile = path.join(tmpDir, `chaos_disk_stress_${process.pid}_${i}.tmp`);
      this.tempFiles.push(tempFile);

      try {
        // Write operation
        const data = Buffer.alloc(fileSizeMB * 1024 * 1024);
        fs.writeFileSync(tempFile, data);

        // Read operation
        fs.readFileSync(tempFile);
      } catch (error) {
        console.warn(`Error during disk I/O: ${error}`);
      }
    }

    // Keep doing I/O operations until duration expires
    const ioInterval = setInterval(() => {
      const elapsed = (Date.now() - startTime) / 1000;
      
      if (elapsed >= duration) {
        clearInterval(ioInterval);
        this.cleanupDiskStress();
        return;
      }

      // Perform read/write cycles on existing files
      for (const tempFile of this.tempFiles) {
        try {
          if (fs.existsSync(tempFile)) {
            // Read
            fs.readFileSync(tempFile);

            // Write (append)
            const data = Buffer.alloc(1024);
            fs.appendFileSync(tempFile, data);
          }
        } catch (error) {
          console.warn(`Error during disk I/O cycle: ${error}`);
        }
      }
    }, 100); // Small delay between cycles
  }

  /**
   * Cleanup disk stress resources.
   */
  private cleanupDiskStress(): void {
    // Cleanup temp files
    for (const tempFile of this.tempFiles) {
      try {
        if (fs.existsSync(tempFile)) {
          fs.unlinkSync(tempFile);
        }
      } catch (error) {
        console.warn(`Error removing temp file ${tempFile}: ${error}`);
      }
    }

    this.tempFiles = [];
    this.activeSimulations.set('disk_stress', false);
    console.log('🔥 CHAOS: Disk I/O stress simulation completed');
  }
}

// Global instance
export const chaosSimulator = new ChaosSimulator();

File: catalog-service/src/chaos_simulator.py
"""Chaos Engineering Simulator for safe infrastructure failure simulation."""

import asyncio
import logging
import os
import random
import threading
import time
from datetime import datetime
from typing import Dict, Any, Optional

try:
    import psutil
except ImportError:
    psutil = None

from .feature_flag_client import FeatureFlagClient

logger = logging.getLogger(__name__)


class ChaosSimulator:
    """Simulates infrastructure failures safely for chaos engineering demos."""
    
    def __init__(self):
        """Initialize the chaos simulator with state tracking."""
        self.active_simulations: Dict[str, bool] = {}
        self.disk_stress_thread: Optional[threading.Thread] = None
        self.crash_scheduler_thread: Optional[threading.Thread] = None
        self.stop_disk_stress = threading.Event()
        self.stop_crash_scheduler = threading.Event()
        
        # Track temp files for cleanup
        self.temp_files: list = []
        
        logger.info("ChaosSimulator initialized")
    
    async def check_and_apply_chaos(
        self, 
        service_name: str, 
        flag_client: FeatureFlagClient
    ) -> None:
        """
        Check for active chaos flags and apply corresponding simulations.
        
        Args:
            service_name: Name of the service (for logging)
            flag_client: Feature flag client to query flags
        """
        try:
            # Check disk stress flag
            disk_stress_flag = await flag_client.get_flag("infrastructure_disk_stress")
            if disk_stress_flag and disk_stress_flag.get("enabled"):
                await self._simulate_disk_stress(disk_stress_flag.get("config", {}))
            
            # Check pod crash flag
            pod_crash_flag = await flag_client.get_flag("infrastructure_pod_crash")
            if pod_crash_flag and pod_crash_flag.get("enabled"):
                await self._simulate_pod_crash(
                    pod_crash_flag.get("config", {}), 
                    service_name
                )
            
            # Check DB connection failure flag
            db_fail_flag = await flag_client.get_flag("infrastructure_db_connection_fail")
            if db_fail_flag and db_fail_flag.get("enabled"):
                await self._simulate_db_connection_fail(db_fail_flag.get("config", {}))
            
            # Check network delay flag
            network_delay_flag = await flag_client.get_flag("infrastructure_network_delay")
            if network_delay_flag and network_delay_flag.get("enabled"):
                await self._simulate_network_delay(network_delay_flag.get("config", {}))
                
        except Exception as e:
            # Don't let chaos simulator errors crash the service
            logger.error(f"Error in chaos simulation: {e}")
    
    async def _simulate_network_delay(self, config: Dict[str, Any]) -> None:
        """
        Simulate network latency between services.
        
        Args:
            config: Configuration with delayMs (base delay) and jitterMs (random variation)
        """
        delay_ms = config.get("delayMs", 2000)
        jitter_ms = config.get("jitterMs", 500)
        
        # Calculate actual delay with random jitter
        jitter = random.randint(-jitter_ms, jitter_ms)
        actual_delay_ms = max(0, delay_ms + jitter)
        
        logger.info(f"🔥 CHAOS: Simulating network delay ({actual_delay_ms}ms)")
        
        # Apply the delay
        await asyncio.sleep(actual_delay_ms / 1000.0)
    
    async def _simulate_db_connection_fail(self, config: Dict[str, Any]) -> None:
        """
        Simulate database connection failures.
        
        Args:
            config: Configuration with failureRate (0-100) and timeoutMs
            
        Raises:
            Exception: When simulating a connection failure
        """
        failure_rate = config.get("failureRate", 50)
        timeout_ms = config.get("timeoutMs", 1000)
        
        # Check if we should fail based on failure rate
        if random.randint(1, 100) <= failure_rate:
            logger.warning(
                f"🔥 CHAOS: Simulating database connection failure "
                f"(rate={failure_rate}%, timeout={timeout_ms}ms)"
            )
            
            # Simulate timeout delay
            await asyncio.sleep(timeout_ms / 1000.0)
            
            # Raise connection exception
            raise Exception("CHAOS: Simulated database connection failure")
    
    async def _simulate_pod_crash(
        self, 
        config: Dict[str, Any], 
        service_name: str
    ) -> None:
        """
        Simulate pod crashes by scheduling periodic exits.
        
        Args:
            config: Configuration with crashIntervalMinutes and crashProbability (0-100)
            service_name: Name of the service for logging
        """
        # Check if already running
        if self.active_simulations.get("pod_crash"):
            return
        
        interval_minutes = config.get("crashIntervalMinutes", 5)
        crash_probability = config.get("crashProbability", 30)
        
        logger.info(
            f"🔥 CHAOS: Starting pod crash scheduler "
            f"(interval={interval_minutes}min, probability={crash_probability}%)"
        )
        self.active_simulations["pod_crash"] = True
        
        # Reset stop event
        self.stop_crash_scheduler.clear()
        
        # Start crash scheduler in background thread
        def crash_scheduler_worker():
            try:
                interval_seconds = interval_minutes * 60
                
                while not self.stop_crash_scheduler.is_set():
                    # Wait for the interval
                    if self.stop_crash_scheduler.wait(timeout=interval_seconds):
                        break  # Stop event was set
                    
                    # Check if we should crash based on probability
                    if random.randint(1, 100) <= crash_probability:
                        logger.error(
                            f"🔥 CHAOS: Simulating pod crash for {service_name} "
                            f"(probability={crash_probability}%)"
                        )
                        # Force exit the process
                        os._exit(1)
                
            finally:
                self.active_simulations["pod_crash"] = False
                logger.info("🔥 CHAOS: Pod crash scheduler stopped")
        
        # Start thread if not already running
        if not self.crash_scheduler_thread or not self.crash_scheduler_thread.is_alive():
            self.crash_scheduler_thread = threading.Thread(
                target=crash_scheduler_worker, 
                daemon=True
            )
            self.crash_scheduler_thread.start()
    
    async def _simulate_disk_stress(self, config: Dict[str, Any]) -> None:
        """
        Simulate high disk I/O by creating intensive read/write operations.
        
        Args:
            config: Configuration with intensityLevel (1-10) and durationSeconds
        """
        # Check if already running
        if self.active_simulations.get("disk_stress"):
            return
        
        intensity = config.get("intensityLevel", 5)
        duration = config.get("durationSeconds", 30)
        
        logger.info(f"🔥 CHAOS: Starting disk I/O stress simulation (intensity={intensity}, duration={duration}s)")
        self.active_simulations["disk_stress"] = True
        
        # Reset stop event
        self.stop_disk_stress.clear()
        
        # Start disk stress in background thread
        def disk_stress_worker():
            try:
                start_time = time.time()
                file_size_mb = intensity  # 1-10 MB per file
                num_files = intensity  # 1-10 files
                
                # Create temp files
                for i in range(num_files):
                    if self.stop_disk_stress.is_set():
                        break
                    
                    temp_file = f"/tmp/chaos_disk_stress_{os.getpid()}_{i}.tmp"
                    self.temp_files.append(temp_file)
                    
                    try:
                        # Write operation
                        with open(temp_file, 'wb') as f:
                            data = os.urandom(file_size_mb * 1024 * 1024)
                            f.write(data)
                            f.flush()
                            os.fsync(f.fileno())
                        
                        # Read operation
                        with open(temp_file, 'rb') as f:
                            _ = f.read()
                    except Exception as e:
                        logger.warning(f"Error during disk I/O: {e}")
                
                # Keep doing I/O operations until duration expires
                while time.time() - start_time < duration:
                    if self.stop_disk_stress.is_set():
                        break
                    
                    # Perform read/write cycles on existing files
                    for temp_file in self.temp_files:
                        if self.stop_disk_stress.is_set():
                            break
                        
                        try:
                            if os.path.exists(temp_file):
                                # Read
                                with open(temp_file, 'rb') as f:
                                    _ = f.read()
                                
                                # Write
                                with open(temp_file, 'ab') as f:
                                    f.write(os.urandom(1024))
                                    f.flush()
                        except Exception as e:
                            logger.warning(f"Error during disk I/O cycle: {e}")
                    
                    time.sleep(0.1)  # Small delay between cycles
                
            finally:
                # Cleanup temp files
                for temp_file in self.temp_files:
                    try:
                        if os.path.exists(temp_file):
                            os.remove(temp_file)
                    except Exception as e:
                        logger.warning(f"Error removing temp file {temp_file}: {e}")
                
                self.temp_files.clear()
                self.active_simulations["disk_stress"] = False
                logger.info("🔥 CHAOS: Disk I/O stress simulation completed")
        
        # Start thread if not already running
        if not self.disk_stress_thread or not self.disk_stress_thread.is_alive():
            self.disk_stress_thread = threading.Thread(target=disk_stress_worker, daemon=True)
            self.disk_stress_thread.start()
    
    def get_system_metrics(self) -> Dict[str, Any]:
        """
        Get current system metrics for monitoring chaos effects.
        
        Returns:
            Dictionary containing system metrics and active simulations
        """
        metrics = {
            "timestamp": datetime.utcnow().isoformat(),
            "active_simulations": [
                name for name, active in self.active_simulations.items() if active
            ]
        }
        
        # Add system metrics if psutil is available
        if psutil:
            try:
                process = psutil.Process()
                metrics["cpu_usage_percent"] = process.cpu_percent(interval=0.1)
                metrics["memory_usage_mb"] = process.memory_info().rss / 1024 / 1024
                
                # Disk I/O metrics
                io_counters = process.io_counters()
                metrics["disk_io_read_bytes"] = io_counters.read_bytes
                metrics["disk_io_write_bytes"] = io_counters.write_bytes
            except Exception as e:
                logger.warning(f"Error collecting system metrics: {e}")
        else:
            logger.warning("psutil not available - system metrics disabled")
        
        return metrics


# Global instance
chaos_simulator = ChaosSimulator()
